src/Umbraco.AuthorizedServices/Services/Implement/AuthorizedServiceCaller.cs

```diff
@@ -229,28 +232,49 @@ internal sealed class AuthorizedServiceCaller : AuthorizedServiceBase, IAuthoriz
+                case AuthenticationMethod.OAuth2AuthorizationCode:
+
+                    // For OAuth2 authorization code flow we can get a new access token if we have a refresh token.
+                    if (string.IsNullOrEmpty(token.RefreshToken))
+                    {
+                        ClearOAuth2AccessToken(serviceAlias);
+                        throw new AuthorizedServiceException($"Cannot request service '{serviceAlias}' as the access token has or will expire and no refresh token is available to use. The expired token has been deleted.");
+                    }
+
+                    return await RefreshAccessToken(serviceAlias, serviceDetail, token.RefreshToken)
+                        ?? throw new AuthorizedServiceException($"Cannot request service '{serviceAlias}' as the access token has or will expired and the refresh token could not be used to obtain a new access token.");
+                case AuthenticationMethod.OAuth2ClientCredentials:
+
+                    // For OAuth2 authorization code flow we can request a new access token via the same means as retrieving the original one.
+                    AuthorizationResult result = await _authorizedServiceAuthorizer.AuthorizeOAuth2ClientCredentialsServiceAsync(serviceAlias);
+                    if (result.Success)
+                    {
+                        OAuth2Token? newToken = OAuth2TokenStorage.GetToken(serviceAlias);
+                        if (newToken is not null)
+                        {
+                            return newToken;
+                        }
+
+                        throw new AuthorizedServiceException($"Cannot request service '{serviceAlias}' as the access token has or will expired and although the client credentials completed successfully a new access token could not be retrieved from storage.");
+                    }
+
+                    throw new AuthorizedServiceException($"Cannot request service '{serviceAlias}' as the access token has or will expired and the client credentials flow did not complete successfully in obtaining a new access token. {result.ErrorMessage}.");
+                default:
+                    throw new ArgumentException($"{serviceDetail.AuthenticationMethod} is not expected here. Only OAuth2 access tokens can be refreshed or renewed.");
             }
-
-            return await RefreshAccessToken(serviceAlias, token.RefreshToken)
-                ?? throw new AuthorizedServiceException($"Cannot request service '{serviceAlias}' as the access token has or will expired and the refresh token could not be used to obtain a new access token.");
         }
 
         return token;
     }
 
-    private async Task<OAuth2Token?> RefreshAccessToken(string serviceAlias, string refreshToken)
+    private async Task<OAuth2Token?> RefreshAccessToken(string serviceAlias, ServiceDetail serviceDetail, string refreshToken)
```

Comment:
Do we still need to pass the `serviceAlias` since we could just pick it from `serviceDetail.Alias`?